katcp/server.py

```diff
@@ -1071,16 +1075,21 @@ def async_reply(f):
                         finally:
                             done_future.set_result(None)
 
-                    # TODO When using the return_reply() decorator the future
-                    # returned is not currently threadsafe, must either deal
-                    # with it here, or in kattypes.py. Would be nice if we don't
-                    # have to always fall back to adding a callback, or wrapping
-                    # a thread-safe future. Supporting sync-with-thread and
-                    # async futures is turning out to be a pain in the ass ;)
-                    self.ioloop.add_callback(reply.add_done_callback, async_reply)
-                    # reply.add_done_callback(async_reply)
-                    self._logger.debug("%s FUTURE OK" % (msg.name,))
-                    return done_future
+                    if getattr(handler, 'concurrent_reply', False):
+                        # Return immediately if this is a concurrent handler
+                        self._logger.debug("%s FUTURE CONCURRENT", msg.name)
+                        return
```

Comment:
I haven't tried running it yet, but this neglects to schedule the call to async_reply.